Harvard Caselaw Access Project.js

```diff
@@ -0,0 +1,175 @@
+{
+	"translatorID": "2a1cafb9-6f61-48d3-b621-c3265fde9eba",
+	"label": "Harvard Caselaw Access Project",
+	"creator": "Franklin Pezzuti Dyer",
+	"target": "^https://(.*\\.)*case\\.law",
+	"minVersion": "5.0",
+	"maxVersion": "",
+	"priority": 100,
+	"inRepository": true,
+	"translatorType": 4,
+	"browserSupport": "gcsibv",
+	"lastUpdated": "2024-01-15 23:41:24"
+}
+
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2024 Franklin Pezzuti Dyer
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+function detectWeb(doc, _url) {
+	if (doc.querySelectorAll('.case-container').length > 0) {
+		return 'case';
+	}
+	else if (getSearchResults(doc, true)) {
+		return 'multiple';
+	}
+	return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('div.result-title > div > a');
+	for (let row of rows) {
+		let href = row.href;
+		Z.debug(href);
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+async function doWeb(doc, url) {
+	if (detectWeb(doc, url) == 'multiple') {
+		let items = await Zotero.selectItems(getSearchResults(doc, false));
+		if (!items) return;
+		for (let url of Object.keys(items)) {
+			await scrape(await requestDocument(url));
+		}
+	}
+	else {
+		await scrape(doc, url);
+	}
+}
+
+async function scrape(doc, url = doc.location.href) {
+	let apiUrl = attr(doc, "a[href*='api.case.law/v1/cases/']", 'href');
+	let caseText = await requestText(apiUrl);
+	let caseJson = JSON.parse(caseText);
+
+	let caseItem = new Zotero.Item("case");
+	caseItem.language = "en-US";
+	caseItem.url = url;
```

Comment:
I left the URL out of the CourtListener translator, but could go either way on this -- do you have any opinion on this? We definitely will want to save an attachment, though -- either a snapshot by using `document: doc` or a PDF. Again, open for either, but we'll definitely want to capture some full text.

Reply:
Went ahead and added the PDF attachment. Do you see any harm in keeping both the case URL and the full text PDF URL? I like redundancy :-)

Reply:
Yeah, I think I'll add it to CourtListener.

For context: Nothing wrong with redundancy -- a URL in the URL field should generally be citeable, though (otherwise it goes into an attached link) -- for academic works, we use it only when it goes to the full text of the work, e.g. For legal materials I think the criteria might be a little different, but I'm not sure -- most people will cite cases without URL, but I'm inclined to see where they are adding the URL, it should be the one they actually retrieved the material from.